enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,385 @@
+pipelines:
+- inputRefs: [ patchyIn, bulldogIn ]
+  outputRefs: default
+```
+
+**Note**: [Equality and set-based label selectors][labels] are supported, example shows only equality based selectors.
+
+#### I want to forward only well-formed JSON with an identified format
+
+**Note**: this requires implementation of [forwarder label selectors][clf-labels]
+
+```yaml
+inputs:
+- name: patchyIn
+  applications:
+    selector:
+      matchLabels: { app: patchy }
+  structured:
+    format: patchy
+outputs:
+- name: patchyOut
+  ...
+  structured:
+    discard: Unstructured
+pipelines:
+- inputRefs: [ patchyIn ]
+  outputRefs: patchyOut
+```
+
+**Note**: [Equality and set-based label selectors][labels] are supported, example shows only equality based selectors.
+
+#### I want to access fields in all JSON log messages.
+
+I am using `fluentForward` and I want to be able to access any field in my JSON log messages.
+I used to use the [`MERGE_JSON_LOG`][merge] feature to do this, but that was removed.
+
+Example 1: I want non-JSON application messages to be discarded. JSON messages are forwarded in the `structured_message` field.
+```yaml
+outputs:
+- name: jsonOut
+  ...
+  structured:
+    discard: unstructured
+pipelines:
+- inputRefs: [ application ]
+  outputRefs: [ jsonOut ]
+```
+
+Example 2: I want both JSON and non-JSON messages forwarded. JSON in the `structured_message` field, non-JSON in the `message` field.
+```yaml
+outputs:
+- name: jsonOut
+  ...
+  structured:
+    discard: none
+pipelines:
+- inputRefs: [ application ]
+  outputRefs: [ jsonOut ]
+```
+
+#### I want to  forward structured and unstructured logs from the same input to different outputs
+
+```yaml
+outputs:
+- name: structuredOut
+  ...
+  structured:
+    discard: unstructured
+- name: unstructuredOut
+  ...
+  structured:
+    discard: structured
+pipelines:
+- inputRefs: [ application ]
+  outputRefs: [ structuredOut, unstructuredOut ]
+```
+
+### Proposal details
+
+#### Data model changes
+
+Two new top-level fields added to the internal [data model][model]:
+* `structured_message` (type object): Original structured JSON log message.
+* `structured_format` (type string): Format name from `input.structured.format`.
+
+Note: `structured_format` is an arbitrary name assigned by the user, it has no meaning to the forwarder.
+
+The following rules apply to forwarded records:
+* At least one of `message` or `structured_message` MUST be present. They MAY both be present.
+* `message` and `structured_message` MAY both be present, if so `message` MUST be the JSON quoted-string form of the value of `structured_message`.
+* `structured_format` MAY be present if `structured_message` is present, and MUST NOT be present if `structured_message` is missing.
+
+Interpretation of the record:
+* `structured_format` is present: it indicates the format of `structured_message`
+* `structured_format` is missing:
+  * `structured_message` is present it: it is JSON, but of an unknown format.
```

Comment:
I wonder if we should choose a different word then "format" like "schema".   JSON is the format. what you are proposing here is for user's that have their own schemas effectively: a named definition  of keys/values that represent a specific service.  we arent asking them to provide the specific schema but a schema name

Reply:
filtering: I see your point, will have a think about it. I was thinking about this more as "routing" than "filtering" - unstructured and structured messages can go different places, including nowhere for unstructured messages if they can't be handled. I've had trouble deciding where to place the controls, I'll reconsider with your comment in mind.

Also I've been round the block on format vs schema vs encoding vs. data model. "Format" appears to be the most popular word used to describe the rules for constructing a log message, but in the case of structured logs that has two parts a) it's a JSON object and b) there are rules for allowed key/value pairs. Maybe 'encoding' and 'data model' and avoid format? E.g. Json and rsyslog are encodings, allowed key-type pairs within those encodings are the data model?

Reply:
See latest, your comments pushed me in a better direction I think.